backends/ref/ceed-ref-restriction.c

```diff
@@ -91,7 +91,8 @@ static inline int CeedElemRestrictionApply_Ref_Core(CeedElemRestriction r,
           CeedPragmaSIMD
           for (CeedInt i = 0; i < elem_size*blk_size; i++)
             vv[elem_size*(k*blk_size+num_comp*e) + i - v_offset]
-              = uu[impl->offsets[i+elem_size*e] + k*comp_stride];
+              = uu[impl->offsets[i+elem_size*e] + k*comp_stride] *
+                (impl->orient && impl->orient[i+elem_size*e] ? -1. : 1.);
```

Comment:
I hope this is optimized to lift this conditional outside the loop. Can you confirm that there is no performance impact to the BPs before/after this change?

Reply:
Hmm, It seems it has.
In `main`, with `./bps -problem bp1 -degree 3`
```
  Performance:
    Pointwise Error (max)              : 2.605215e-03
    CG Solve Time                      : 0.00407007 (0.00407007) sec
    DoFs/Sec in CG                     : 3.68544 (3.68544) million
 ```
 In this `branch`, with `./bps -problem bp1 -degree 3`
 ```
   Performance:
    Pointwise Error (max)              : 2.605215e-03
    CG Solve Time                      : 0.00250097 (0.00250097) sec
    DoFs/Sec in CG                     : 5.99767 (5.99767) million
 ```

Reply:
For other problem as well, 
In `main`, with `./bps -problem bp4 -degree 3`
```
  Performance:
    Pointwise Error (max)              : 1.273306e-01
    CG Solve Time                      : 0.00880826 (0.00880826) sec
    DoFs/Sec in CG                     : 1.56944 (1.56944) million
```
here
```
  Performance:
    Pointwise Error (max)              : 1.273306e-01
    CG Solve Time                      : 0.00512111 (0.00512111) sec
    DoFs/Sec in CG                     : 2.69941 (2.69941) million
```

Reply:
maybe we can add something of `orient` exist or not?!

Reply:
You can make `is_oriented` as a const bool input. That might be enough, if not, you can make versions like we have for dimensions.

Reply:
ok, thanks.

Reply:
I'm very skeptical of those results (which would say it's a lot faster here), but also very slow. What optimization flags?
```console
$ build/petsc-bps -degree 3 -local_nodes 100000

-- CEED Benchmark Problem 1 -- libCEED + PETSc --
  MPI:
    Hostname                           : kichatna
    Total ranks                        : 1
    Ranks per compute node             : 1
  PETSc:
    PETSc Vec Type                     : seq
  libCEED:
    libCEED Backend                    : /cpu/self/xsmm/blocked
    libCEED Backend MemType            : host
  Mesh:
    Number of 1D Basis Nodes (P)       : 4
    Number of 1D Quadrature Points (Q) : 5
    Global nodes                       : 106720
    Local Elements                     : 3705
    Owned nodes                        : 106720
    DoF per node                       : 1
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 13
    Final rnorm                        : 8.283258e-11
  Performance:
    Pointwise Error (max)              : 5.122931e-04
    CG Solve Time                      : 0.0282211 (0.0282211) sec
    DoFs/Sec in CG                     : 49.1605 (49.1605) million
```

Reply:
I build using `make OPT="-O3 -march=native" -B` and now with -localnodes 100000 I get
```
 ./bps -problem bp1 -degree 3 -local_nodes 100000

-- CEED Benchmark Problem 1 -- libCEED + PETSc --
  MPI:
    Hostname                           : rezgar
    Total ranks                        : 1
    Ranks per compute node             : 1
  PETSc:
    PETSc Vec Type                     : seq
  libCEED:
    libCEED Backend                    : /cpu/self/avx/blocked
    libCEED Backend MemType            : host
  Mesh:
    Number of 1D Basis Nodes (P)       : 4
    Number of 1D Quadrature Points (Q) : 5
    Global nodes                       : 106720
    Local Elements                     : 3705
    Owned nodes                        : 106720
    DoF per node                       : 1
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 13
    Final rnorm                        : 8.282633e-11
  Performance:
    Pointwise Error (max)              : 5.122931e-04
    CG Solve Time                      : 0.0721097 (0.0721097) sec
    DoFs/Sec in CG                     : 19.2396 (19.2396) million
```

Reply:
In `main` I get
```
./bps -problem bp1 -degree 3 -local_nodes 100000

-- CEED Benchmark Problem 1 -- libCEED + PETSc --
  MPI:
    Hostname                           : rezgar
    Total ranks                        : 1
    Ranks per compute node             : 1
  PETSc:
    PETSc Vec Type                     : seq
  libCEED:
    libCEED Backend                    : /cpu/self/avx/blocked
    libCEED Backend MemType            : host
  Mesh:
    Number of 1D Basis Nodes (P)       : 4
    Number of 1D Quadrature Points (Q) : 5
    Global nodes                       : 106720
    Local Elements                     : 3705
    Owned nodes                        : 106720
    DoF per node                       : 1
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 13
    Final rnorm                        : 8.282633e-11
  Performance:
    Pointwise Error (max)              : 5.122931e-04
    CG Solve Time                      : 0.0823446 (0.0823446) sec
    DoFs/Sec in CG                     : 16.8482 (16.8482) million
```

Reply:
Now with `(is_oriented && impl->orient[i + elem_size * e] ? -1. : 1.)`, in `this branch` I get almost same time as `main`.
```
./bps -problem bp1 -degree 3 -local_nodes 100000

-- CEED Benchmark Problem 1 -- libCEED + PETSc --
  MPI:
    Hostname                           : rezgar
    Total ranks                        : 1
    Ranks per compute node             : 1
  PETSc:
    PETSc Vec Type                     : seq
  libCEED:
    libCEED Backend                    : /cpu/self/avx/blocked
    libCEED Backend MemType            : host
  Mesh:
    Number of 1D Basis Nodes (P)       : 4
    Number of 1D Quadrature Points (Q) : 5
    Global nodes                       : 106720
    Local Elements                     : 3705
    Owned nodes                        : 106720
    DoF per node                       : 1
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 13
    Final rnorm                        : 8.282633e-11
  Performance:
    Pointwise Error (max)              : 5.122931e-04
    CG Solve Time                      : 0.0828976 (0.0828976) sec
    DoFs/Sec in CG                     : 16.7358 (16.7358) million
```